hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -489,13 +489,29 @@ public EntityEntry addEntry(
 
 		final EntityEntry e;
 
-		if( (entity instanceof ManagedEntity) &&  ((ManagedEntity) entity).$$_hibernate_getEntityEntry() != null && status == Status.READ_ONLY) {
-			e = ((ManagedEntity) entity).$$_hibernate_getEntityEntry();
-			e.setStatus( status  );
+		/*
+		Important: 		the following instanceof checks and castings are intentional.
+						Please do *not* refactor to make calls through the EntityEntryFactory
+						interface. Refactoring will results in polymorphic call sites which will
+						severely impact performance.
+		*/
+		if (persister.getEntityEntryFactory() instanceof MutableEntityEntryFactory){
+			e = ((MutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
+					status,
```

Comment:
We are checking which EntityEntryfactory is referenced in the persister and casting to the implementation of the EntityEntry Factory. If we make the call via the interface, the call is to a polymorphic call site, which severely impacts performance, making the direct call means the JVM does not have to resolve which implementation to use.